### DISCUSSION TITLE
Serverless deployment

### DISCUSSION THREAD

**Discussion by samehhady** [directus/directus#15572] [Feature Requests]
I am wondering if anyone was able to deploy Directus using Serverless framework or being able to deploy Directus as Lambda function.

This would be very beneficial and speed up the deployment process and cost.

**Comment by thedeadofblackfire** [comment]
Yes at least for the api side. 
I played with Vercel (that use AWS lambda under the hood) and Directus with some extensions. 
Of course, uploads need to be on S3 and not local.

The response time are correct, around 15 - 30 ms per request depends on cache.
You can also disable some database/file checkup inside api/app.ts to speed up the boot time.

For now, I got failures on admin but it should be possible.

**Reply by thedeadofblackfire** [reply]
With Serverless framework & AWS Lambda, you can use this link (https://github.com/AbraamNader/directus-serverless-template) as starting script.

**Reply by samehhady** [reply]
Thanks for the reply, looking good but I guess I will still need to host the CMS part somewhere else. I am more looking to wrap up all Directus environment in a lambda function, similar implementation been done to NextJS https://www.serverless.com/plugins/serverless-nextjs-plugin

**Comment by rijkvanzanten** [comment]
Directus itself isn't designed to be ran natively on Lambda functions at this point. You might be able to get things going by using the exported express app from `directus/app.js` and plug that in some sort of wrapper like the above linked serverless-template is doing, but your mileage may vary. There's multiple pieces in the API that rely on environment features that may or may not be available in those worker runtimes, like persistent connections (db/redis), local (temp) filesystem, and persistent memory. Similarly, seeing that you'd effectively run the whole of Directus as a single lambda function, you might run into some serious performance bottlenecks in cold starts.

**Comment by florianbepunkt** [comment]
We are also interested in using Directus in a serverless env. Our use case is to use directus as a central data input/storage platform for non-technical personel with low-frequency access (other systems, mainly static site generators, would query directus during build time).

I managed to deploy directus via cdk, using SQLite via AWS EFS as persistent storage mechanism. @rijkvanzanten Could you please give a hint how to separate the admin ui? I guess it is a Vue or React app that communicates with the directus api, correct? If so it should be possible to separate the admin ui and deploy it to CloudFront.

**Reply by rijkvanzanten** [reply]
This is not something we have tried or done before, so I have no idea what the correct way is. That being said, the app static bundle is distributed through `@directus/app`, which is what the api normally serves under `/admin`:

https://github.com/directus/directus/blob/f940b6a1fce9959c03500aafb26aba4221bb8cb1/api/src/app.ts#L181-L182

The app does require to be under the API domain under path `/admin`, so keep that in mind in your deployment strategy

**Reply by florianbepunkt** [reply]
Thank you. It is possible to decouple the API and the frontend if I set  API_URL env (for dev mode) 

https://github.com/directus/directus/blob/f940b6a1fce9959c03500aafb26aba4221bb8cb1/app/vite.config.js#L144

or change the util function for a production build

https://github.com/directus/directus/blob/f940b6a1fce9959c03500aafb26aba4221bb8cb1/app/src/utils/get-root-path.ts#L4

It would be great if this could be configured without resorting to a custom build.



As far as I can tell, there is no substantial obstacle to deploying Directus to a serverless environment. It probably remains a niche case though.

**Reply by rijkvanzanten** [reply]
I don't know! You've reached uncharted territories. Please do report back with your findings though! If it does turn out to be easy to run Directus in this fashion, I'm down to discuss supporting it in a more official capacity 🙂

**Comment by florianbepunkt** [comment]
@rijkvanzanten

I dropped the SQLite approach (it works, but leads to write conflicts pretty quickly) and used Cockroach Serverless DB instead. Here are my findings so far. There are some things that are not directy related to serverless infrastructure.

## Lambda handler

This is my handler.js. I use a docker image for lambda to circumvent the deployment package size limit.

```js
import directus from "directus";
import express from "@vendia/serverless-express";

const { createApp } = directus;

let instance = null;

const setup = async (event, context) => {
  const app = await createApp();
  instance = express({ app });
  return instance(event, context);
};

export const handler = async (event, context) => {
  if (instance) return instance(event, context);
  return setup(event, context);
};
```

## Admin frontend

Currently the admin frontend is tighliy coupled to the API from a deployment perspective. This is not related to serverless deployments only. It might be beneficial to allow the user to deploy the frontend decoupled from the API server for performance and/or security reasons (e. g. give access to the frontend within company's VPC, etc.).

The coupling happens in two places:

### (1) API URL

The api base url is deducted from the domain the frontend is hosted from.

### (2) Authentication

The authentication as it is cookie based. Using another persistence mechanism for the refresh token allows different domains for frontend and backend. It should be noted though, that storing tokens in local storage has its own security implications and it is advisable to use way shorter refresh token expiration when using local storage than with secure cookies.

## Vite bug

I had some relative path issues while deploying the admin app to S3/Cloudfront. It turned out this was an issue with Vite that has been fixed from version 3 onwards. I updated the vite dependencies – error gone.

## Sporadic errors

I had some sporadic errors that I couldn't reliably reproduce. They happen rather frequently, especially if I perform a reload. First I though it to be CORS issue... however on inspecting detailed logs from the api gateway I found, that the initial lambda concurrency quota for this new test account was quite low (10)... and directus was hitting the limit. This took quite some time to figure out – so if someone experiences similiar issues check your quota in the service quota dashboard.

## S3 storage adapter errors

While I can upload to S3 from directus and delete files, I have issues seeing file previews. They do not load. It logs some errors:

```
15:02:17 🚨 The "path" argument must be of type string. Received null
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received null
at new NodeError (node:internal/errors:372:5)
at validateString (node:internal/validators:120:11)
at Object.join (node:path:1172:7)
at AmazonWebServicesS3Storage._fullPath (/var/task/node_modules/@directus/drive-s3/dist/AmazonWebServicesS3Storage.js:39:61)
at AmazonWebServicesS3Storage.exists (/var/task/node_modules/@directus/drive-s3/dist/AmazonWebServicesS3Storage.js:86:25)
at AssetsService.getAsset (/var/task/node_modules/directus/dist/services/assets.js:74:71)
at processTicksAndRejections (node:internal/process/task_queues:96:5)
at async /var/task/node_modules/directus/dist/controllers/assets.js:123:36

```

Found out that `path.join` in mentioned file above joins an optional root path – but by joining them, at least in node 16, the root is not optional... a default value of `"/" ` should be set, as path.join with an undefined root throws an error.

Fixed this error, but still not seing thumbsnails / images in directus admin app.

## Lots of console errors

There are a lot of 404 since directus is trying to open barrel files within the extensions folder, e. g. extensions/interfaces/index.js. However these files do not exist (at least in my project created following the docs). I guess this issue has nothing to do with serverless and would happen for other deployment methods as well, depending on individual server config regarding missing files.

## Performance

There is some cold start time (as expected with lamdba). I cache the directus app within the function handler, so subsequent calls reusing the function environment are blazingly fast. For my current use case (internal data management app, small team) this is not an issue.

## My current deployment workflow

I forked the directus repo, made beforementioned changed to vite dependencies and auth token handling. Then I deploy my lambda handler as well as s3 buckets, Cloudfront distribution, EFS storage (for extensions) and API gateway via CDK. The biggest obstacle is the changed that need to be made to the directus app. Allowing a decupled frontend as an option would help a lot.

**Reply by rijkvanzanten** [reply]
Impressive stuff!

**Reply by peki1** [reply]
Hey @florianbepunkt can you please provide sample Dockerfile and serverless.yml, we would like to try it too

**Reply by BrianJM** [reply]
> Hey @florianbepunkt can you please provide sample Dockerfile and serverless.yml, we would like to try it too

@peki1 Try forking [AbraamNader/directus-serverless-template](https://github.com/AbraamNader/directus-serverless-template) and applying the `handler.js` changes [here](https://github.com/directus/directus/discussions/15572#discussioncomment-3707336).

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by BrianJM** [reply]
Added my vote just now.

**Reply by aguiraf** [reply]
here is my vote!

**Reply by Swahjak** [reply]
He @rijkvanzanten, since this has now reached 23 votes, would you reconsider this?

**Reply by rijkvanzanten** [reply]
Sure!

**Comment by emilienbidet** [comment]
It would be so cool to be able to deploy a directus instance on Vercel. I think it would help the mass adoption. Deploying the front and the backend on Vercel, whaaat a dev experience !! <3

**Comment by drobertson-dev** [comment]
Yes please add this, the docker image is great, and I can throw on a digital ocean droplet very easily, but hosting Directus and the front facing website at the same company, Vercel would be pretty great.

**Comment by julian-hecker** [comment]
Would love to see this functionality. I can barely justify the cost of an EC2 instance if I'm just using this as a CMS for static sites which would get hit once at build time.

**Reply by rijkvanzanten** [reply]
> if I'm just using this as a CMS for static sites which would get hit once at build time.

By using it as a CMS wouldn't you inherently use the app to manage the content in the content management system and therefore use the platform? I'm actually curious if a serverless deployment like Lambda with an API Gateway would end up being cheaper than leaving a cheap ec2 box running for that use case 🤔

**Reply by rijkvanzanten** [reply]
Especially when we start taking things like caching and cold-start performance into account 🤔🤔 We don't have the research here yet of course, but I'd actually be surprised if a serverless architecture would be a more economical hosting option for a full CMS.

**Reply by thedeadofblackfire** [reply]
Sure, serverless lambda cut us the server instances bill per 2 or 3 compared to permanent ec2. 

For a php project for example, lambda with bref replaces the 3 apache and 1 load balancing dedicated.

For my experience, lambda has a tmp folder with writable permissions to save local filesystem cache.

A lambda instance can be share with multiple users/browsers during some minutes. The max duration i see in live monitoring was 15min.
During this time, the cache saved in tmp are available for all users on the same lambda instance. The lambda instance can stop at anytime. 

Depends on application, but node js with express on lambda should have the same benefits.

**Reply by rijkvanzanten** [reply]
To your own point:

> Depends on application

I am not comfortable making the claim that running Directus completely serverless will by definition cut costs by 2x or 3x, or even any at all.

**Reply by thedeadofblackfire** [reply]
Yes my last sentence has some precautions. I have more experience in lambda "php" using directly AWS or using VERCEL. 

I had tested a prototype of directus "api" (v9) port on vercel serverless. I had to rewrite part of the file access of the database\system-data\fields and at the storage level. The api worked very well. The api benchmark & cold start performance were there !

The problem was mainly around the admin interface, I had to do too many tweaks on js bundle & paths and therefore I deviated too much from the original code.

For me, this “serverless” feature is a success. The CMS part (I mean ADMIN) should be encapsulated in a standalone project (using Vue encapsulation in Nuxt) and leave for example the API and CMS deployments on VERCEL in 2 separate projects. This also makes sense on the “security” side for administrators and content managers for the CMS part, isolated from the API infrastructure. It's possible with keeping source into the current mono repository.

More than cost reduction, the "One-Click Deploy" button for vercel or any other platform should help our community to test Directus easily and set up prototypes "effortlessly" at the infra level (before "maybe " switch to a more robust instance in production depending on the volume and benefits of node sharding via pm2 self hosting or using Directus cloud).

**Comment by rickylabs** [comment]
> Yes my last sentence has some precautions. I have more experience in lambda "php" using directly AWS or using VERCEL.
> 
> I had tested a prototype of directus "api" (v9) port on vercel serverless. I had to rewrite part of the file access of the database\system-data\fields and at the storage level. The api worked very well. The api benchmark & cold start performance were there !
> 
> The problem was mainly around the admin interface, I had to do too many tweaks on js bundle & paths and therefore I deviated too much from the original code.
> 
> For me, this “serverless” feature is a success. The CMS part (I mean ADMIN) should be encapsulated in a standalone project (using Vue encapsulation in Nuxt) and leave for example the API and CMS deployments on VERCEL in 2 separate projects. This also makes sense on the “security” side for administrators and content managers for the CMS part, isolated from the API infrastructure. It's possible with keeping source into the current mono repository.
> 
> More than cost reduction, the "One-Click Deploy" button for vercel or any other platform should help our community to test Directus easily and set up prototypes "effortlessly" at the infra level (before "maybe " switch to a more robust instance in production depending on the volume and benefits of node sharding via pm2 self hosting or using Directus cloud).

Agree with that comment !
+1 for a Nuxt standalone Admin decoupled from backend 
+1 for supporting deployment on platform like Vercel (that also offer severless PG btw) 

Would love to see if Directus team takes on that challenge !

**Comment by stacktasec** [comment]
I'm eager to see this feature implemented. 

After comparing several popular serverless service providers, such as Vercel and CloudFlare, I've noticed that CloudFlare offers a significant cost advantage when it comes to outbound traffic—in fact, it does not price this aspect at all. 

Therefore, I am very hopeful that Directus could be made compatible with CloudFlare Workers, as well as with other serverless providers, enabling it to run on these platforms!

**Comment by GiamPy5** [comment]
You might be interested in ECS Fargate (serverless) deployment maybe! I've developed a Terraform module to easily deploy Directus :) https://github.com/GiamPy5/terraform-aws-directus

**Reply by rijkvanzanten** [reply]
+1! Fargate or other container-based "serverless" deployment paradigms are generally better for executing long-running processes like Directus

**Reply by julian-hecker** [reply]
I think this sounds exactly what I'm looking for!

**Reply by thedeadofblackfire** [reply]
Just brillant ! Do you know if it's full compatible with vercel terraform : https://vercel.com/guides/integrating-terraform-with-vercel ? So to deploy under vercel servers ? I have no knowledge in terraform for the moment ;)

**Reply by samehhady** [reply]
For sure Fargate would be better as a serverless approach yet It is more of a production approach not for quick POC and development as you will have to setup many things for it to work like ECS task, Load balancers, target groups etc which is not as simple as hosting the full thing in S3 and serverless lamba function

**Reply by thedeadofblackfire** [reply]
yep agreed on the "complexity" to configure AWS. Make effort & sense for production environment but not to POC something on Directus. I prefer use Caddyserver on top of multiple instances of directus deployed on a single "low cost" cloud compute server on Vultr to test in a production spirit different "genesis" projects.

**Reply by GiamPy5** [reply]
> For sure Fargate would be better as a serverless approach yet It is more of a production approach not for quick POC and development as you will have to setup many things for it to work like ECS task, Load balancers, target groups etc which is not as simple as hosting the full thing in S3 and serverless lamba function

That's exactly why the module has been developed so that you don't have to think about all of these things on your own! It just works 😁

**Reply by GiamPy5** [reply]
> Just brillant ! Do you know if it's full compatible with vercel terraform : https://vercel.com/guides/integrating-terraform-with-vercel ? So to deploy under vercel servers ? I have no knowledge in terraform for the moment ;)

No, this module only deploys Directus in AWS.
